4456 optional import issue on top-level modules -- monai.handlers/monai.engines

wyli opened the pull request:
Fixes #4456

### Description
make sure that if the top-level modules miss dependencies, the optional import error raises properly.

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [ ] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

wyli commented:
hi @ericspod could you please help review this PR? this is to correct the error message when top-level modules (`monai.engines` and `monai.handlers`) are not imported by `load_modules`, it now raises an optional import error instead of an attribute error. All integration tests work fine.

ericspod commented:
The way of loading modules needs to be reconsidered a bit here since the two stage process was from some very old code of mine. We aren't using the `@export` decorator which it was intended for. The immediate issue here however is that `Metric` object [here](https://github.com/Project-MONAI/MONAI/blob/41950d4a3c6498d7a4e26a8ae08b0fbb2246cad0/monai/handlers/ignite_metric.py#L22) raises the exception when it is used as a parent class, I would say it should raise the exception when one tries to instantiate it instead. Because the exception is raised the module doesn't get created, your solution here doesn't raise the error but will return a `LazyRaise` object which is a bit confusing as well. Could we change the behaviour of `LazyRaise` so that `monai.handlers` is an actual module with classes which raises exceptions on instantiation?

wyli commented:
thanks @ericspod, it seems it's possible to achieve what you suggested. I've added an `as_type` bool to `optional_import`, when it's true, it returns the class instead of an instance so other types can inherit. I've updated the PR, please let me know what you think...

ericspod commented:
The code from issue #4456 now works in that it gives me a module object back as expected, however it permits me to instantiate `monai.handlers.IgniteMetric` without Ignite being installed, it does raise an exception when I try to call any of its methods. If that's expected then it works and I'm good with the changes, I would still suggest some tweak that will raise the exception on instantiation however. This may require a different class from `LazyRaise` which has a constructor accepting any arguments and doing nothing but raising the exception.

wyli commented:
updated to raise an error when calling `__init__`, the only corner case is the decorator usage:

https://github.com/Project-MONAI/MONAI/blob/d64c6d2e88a658ec02d5fd23559513dec24153ff/monai/handlers/ignite_metric.py#L63

it'll also call the constructor, so I add some check for this case. @ericspod

ericspod reviewed (approved):
Works for me as expected now, thanks!

wyli commented:
/build

wyli commented:
/build
